// element-plus/es/hooks/use-form-item/index.js
import { inject, computed, getCurrentInstance, unref } from 'vue';
import { elFormKey, elFormItemKey } from '../../tokens';
import { buildProp } from '../../utils/props';
import { useGlobalConfig } from '../../utils/util';
const sizes = ['', 'large', 'medium', 'small', 'mini'];
export const useFormItemProps = {
    size: buildProp({
        type: String,
        values: sizes,
        default: '',
    }),
    disabled: Boolean,
};
export const useFormItem = ({ size, disabled }) => {
    const vm = getCurrentInstance();
    const $ELEMENT = useGlobalConfig();
    // vm.props is not reactive so we use the reactive one here.
    const props = vm.proxy.$props;
    const form = inject(elFormKey, undefined);
    const formItem = inject(elFormItemKey, undefined);
    return {
        size: computed(() => {
            // TODO, fallback to default size like 'medium/large' instead of empty string
            return (props.size ||
                unref(size) ||
                (formItem === null || formItem === void 0 ? void 0 : formItem.size) ||
                (form === null || form === void 0 ? void 0 : form.size) ||
                $ELEMENT.size ||
                '');
        }),
        disabled: computed(() => {
            return (props.disabled === true || unref(disabled) || (form === null || form === void 0 ? void 0 : form.disabled) || false);
        }),
    };
};
